reference/best-practices/java-app.md

```diff
@@ -0,0 +1,362 @@
+
+### Probe configuration
+
+The connection pool maintains long connections to TiDB. TiDB does not proactively close client connections by default (unless an error is reported), but generally there will be network proxies such as LVS or HAProxy between the client and TiDB. Usually, these proxies will proactively clean up connections that are idle for a certain period of time. In addition to paying attention to the idle configuration of the proxies, the connection pool also needs to keep alive or probe connections.
+
+If you often see the following error in your Java application:
+
+```
+The last packet sent successfully to the server was 3600000 milliseconds ago. The driver has not received any packets from the server. com.mysql.jdbc.exceptions.jdbc4.CommunicationsException: Communications link failure
+```
+
+If `n` in `n milliseconds ago` is `0` or a very small value, it is usually because the executed SQL operation causes TiDB to exit with exception. To solve this, it is recommended to check the TiDB stderr log.
+
+If `n` is a very large value (such as `3600000` in the example), it is likely that this connection was idle for a long time and then closed by the intermediate proxy. The usual solution is to increase the value of the proxy's idle configuration and allow the connection pool to:
+
+- Check whether the connection is available before using the connection every time
+- Periodically check whether the connection is available using a separate thread
+- Send a test query regularly to keep alive connections
+
+Different connection pool implementations may support one or more of the above methods. You can check your connection pool documentation to find the corresponding configuration.
+
+## Data access framework
+
+Applications often use some kind of data access framework to simplify database access.
+
+### MyBatis
+
+[MyBatis](http://www.mybatis.org/mybatis-3/) is currently a popular Java data access framework. It is mainly used to manage SQL queries and complete the mapping of result sets and Java objects. MyBatis is highly compatible with TiDB. MyBatis rarely has problems based on its historical issues.
+
+Here we mainly focus on the following configurations.
+
+#### Mapper parameter
+
+MyBatis Mapper supports two parameters:
+
+- `select 1 from t where id = #{param1}` will be converted to `select 1 from t where id =?` as a Prepared Statement and be "prepared", and the actual parameter will be used for reuse. You can achieve the best performance when using this parameter with the previously mentioned Prepare connection parameters.
+- `select 1 from t where id = ${param2}` will be replaced with `select 1 from t where id = 1` as a text file and be executed. If this statement is "prepared" to different parameters, it may cause TiDB to cache a large number of Prepared Statements, and executing SQL operations this way has injection security risks.
```

Comment:
```suggestion
- `select 1 from t where id = ${param2}` will be replaced with `select 1 from t where id = 1` as a text file and be executed. If this statement is replaced with different parameters and is executed, MyBatis will send different requests for "preparing" the statements to TiDB. This might cause TiDB to cache a large number of Prepared Statements, and executing SQL operations this way has injection security risks.
```